src/Workspaces/Remote/ServiceHub/Host/RemoteWorkspace.cs

```diff
@@ -60,29 +62,164 @@ protected override void Dispose(bool finalize)
+        /// 
+        /// but doing that from scratch all the time wil be expansive in terms of synching data, compilation being cached, file being parsed
+        /// and etc. so even if the service itself is stateless, internally it has several caches to improve perf of various parts.
+        /// 
+        /// first, it holds onto last solution got built. this will take care of common cases where multiple services running off same solution.
+        /// second, it uses assets cache to hold onto data just synched (within 3 min) so that if it requires to build new solution, 
+        ///         it can save some time to re-sync data which might just used by other solution.
+        /// third, it holds onto solution from primary branch from Host. and it will try to see whether it can build new solution off the
+        ///        primary solution it is holding onto. this will make many solution level cache to be re-used.
+        ///
+        /// the primary solution can be updated in 2 ways.
+        /// first, host will keep track of primary solution changes in host, and call OOP to synch to latest time to time.
+        /// second, engine keeps track of whether a certain request is for primary solution or not, and if it is, 
+        ///         it let that request to update primary solution cache to latest.
+        /// 
+        /// these 2 are complimentary to each other. #1 makes OOP's primary solution to be ready for next call (push), #2 makes OOP's primary
+        /// solution be not stale as much as possible. (pull)
         /// </summary>
-        public bool TryAddSolutionIfPossible(SolutionInfo solutionInfo, int workspaceVersion, SerializableOptionSet options, [NotNullWhen(true)] out Solution? solution)
+        private async Task<Solution> CreateSolution_NoLockAsync(
+            AssetProvider assetProvider,
+            Checksum solutionChecksum,
+            bool fromPrimaryBranch,
+            int workspaceVersion,
+            Solution baseSolution,
+            CancellationToken cancellationToken)
         {
-            if (solutionInfo == null)
+            try
             {
-                throw new ArgumentNullException(nameof(solutionInfo));
+                var updater = new SolutionCreator(Services.HostServices, assetProvider, baseSolution, cancellationToken);
+
+                // check whether solution is update to the given base solution
+                if (await updater.IsIncrementalUpdateAsync(solutionChecksum).ConfigureAwait(false))
+                {
+                    // create updated solution off the baseSolution
+                    var solution = await updater.CreateSolutionAsync(solutionChecksum).ConfigureAwait(false);
+
+                    if (fromPrimaryBranch)
+                    {
+                        // if the solutionChecksum is for primary branch, update primary workspace cache with the solution
+                        return UpdateSolutionIfPossible(solution, workspaceVersion);
```

Comment:
don't love taht this mutates here.  would prefer the caller handle this.

Reply:
and if it can mutate, i woudl prefer the name of the method makes that clear.  i.e. CreateSolutionAndPotentiallyUpdatePrimarySolution_NoLockAsync

Reply:
This code was just moved, we can refactor later.